### ISSUE TITLE
[Question] How to enable custom commands autocompletion ?

### ISSUE THREAD

**Issue by tianalemesle** [mbolotov/intellij-cypress#50]
I have multiple `Cypress.Commands.add` calls from `cypress/support/commands.js`, how do I enable autocompletion for those ?

The `Add type for cypress custom command` intention is enabled but I can't find a way to trigger it.

Thanks

**Comment by mbolotov** [comment]
Hi @tianalemesle! 

The most preferred way to get autocompletion is to use TypeScript definitions for the commands. 
I think the intention you mentioned is not working for you because by default it try to generate the definitions in the same file where the command itself is placed. 
But as you have a 'js' file for you command, it fails to do this. It should be a 'ts' one. 
I think the plugin UI should be improved for such a case. 
Currently, you have 2 options to set the definition file path. 
1. If you run IDEA, you can access this setting in the cypress facet page.
![image](https://user-images.githubusercontent.com/19785715/123942908-41fe9780-d9a4-11eb-8634-abe1a8f6a66f.png)

2. If you any other IntelliJ IDE (like WebStorm), it probably has no 'module' notion at all and thus no facet either. But you can still define the file path using 'Extract cypress command' action:
![image](https://user-images.githubusercontent.com/19785715/123943151-87bb6000-d9a4-11eb-899a-67f6d143eaa3.png)
![image](https://user-images.githubusercontent.com/19785715/123943395-c9e4a180-d9a4-11eb-9559-59472d069b3c.png)
You have to just select the file name and click 'Cancel' - it will save the default value even if no extract action is performed. 

I will probably rework the UI to make it easier to select the definition file.

**Comment by tianalemesle** [comment]
I see, however, I'm not using TypeScript at all...

**Comment by mbolotov** [comment]
You probably do not need to add any support for TypeScript into your project. 
It's for IDE only, just try to add this single ts file.

**Comment by mbolotov** [comment]
Hi @tianalemesle!
Have had a chance to try adding the definition file?

BTW, I just realized that the current version of the plugin should suggest autocompletion for a command without a type definition. But this is for code inside a java script file only (not a ts one).
I tried on this project: https://github.com/rightsaidjames/cypress-starter-kit.git
And it worked okay:
![image](https://user-images.githubusercontent.com/19785715/124973362-1a5a9f80-e034-11eb-98ee-4c08840acf1d.png)

**Comment by tianalemesle** [comment]
If that TypeScript file needs to be located in the project directory, then I won't.

Yes, I realized autocompletion actually works.

However, I still get the `Method expression is not of Function type` warning.

**Comment by mbolotov** [comment]
> Method expression is not of Function type
I'll look if this could be fixed.
But I doubt as IDE does not really know your command unless it is defined in a TS file.

**Comment by tianalemesle** [comment]
How can the command be autocompleted but not highlighted ?

**Comment by mbolotov** [comment]
What do you mean by 'not highlighted'?
IntelliJ API allows to add custom completion contributor, reference resolver and other extensions.
The problem is that there is a default JS code analyzer that has no knowledge about custom commands and thus it warns about 'unexpected' method calls. I'm not sure if the plugin could somehow change this default behavior.

**Comment by tianalemesle** [comment]
Here I have a route named `foo-loaded` and a command named `any`.

Both are autocompleted and both have the `Ctrl+B` (`Go to declaration`) shortcut working.
I would think this means the analyzer actually knows the definition of both.
Yet, the route is highlighted while the command is not.

![Screenshot from 2021-07-09 14-27-22](https://user-images.githubusercontent.com/72209627/125077791-ca2a1e80-e0c1-11eb-85c0-c89373820250.png)

**Comment by mbolotov** [comment]
Aliases and custom methods are not the same things and are handled differentially.
Alias handling is fully implemented on the plugin side. From the core IDE perspective, it's just a string literal. 
On the contrary, a custom command is not a literal and IDE expect it to be a valid JS method call. 
This is why it's complaining about a custom method and not about an alias.

**Comment by tianalemesle** [comment]
But, if I press `Ctrl+B` on `any()`, it leads me to the `Cypress.Commands.add` declaration, how can it do that and yet display as it was unrecognized ?

**Comment by mbolotov** [comment]
It's just because the plugin adds a reference to the declaration. 
But it's not enough for the default JS analyzer. It expects that a method call go to a valid JS declaration. But custom command is just a string literal, not a JS declaration. 
I think this issue could probably done in the same way as it is for aliases (by injecting plugin own parser and PSI structure inside the string literal). 

That being said, it requires a lot of work and I'm not sure if this feature is in demand enough.

**Comment by mbolotov** [comment]
BTW, can you explain why it's not possible to add a single TS file inside your project?

**Comment by tianalemesle** [comment]
Because I'm the only one among my coworkers who uses your plugin and will subscribe.

**Comment by mbolotov** [comment]
There are should be really a lot of coworkers to fund this feature development.
Anyway, I still don't understand the problem of having a TS file in your project. If you can explain it to me, it could probably change my mind in the terms of demand estimation.

**Comment by tianalemesle** [comment]
I don't think it will, the reason is, we're collaborating on a project, which uses Cypress for tests, and I can't throw a file on the repo that is useless to everyone but me.

**Comment by mbolotov** [comment]
This file structures your tests and makes them more robust and clean. So I would say that this file is useful for anyone who look into the tests.

**Comment by MacGyver27** [comment]
None of this worked for me to get syntax highlighting.. neither custom commands recognition. I am evaluating the pro version at the moment.

Support folder looks like this:
![image](https://user-images.githubusercontent.com/6546231/131342966-51579cd6-3293-44e6-a6e6-c90d8e3f489a.png)

Also tried to rename `commands.js` to `commands.ts` - didnt help. Did that for `index.js` also noting..

Tried also installing `cypress-intellij-reporter` but not sure what it is for..

![2021-08-30_15-04-12](https://user-images.githubusercontent.com/6546231/131343393-9f7828f5-ec8b-4971-9233-681379764972.gif)

**Comment by mbolotov** [comment]
> None of this worked for me to get syntax highlighting.. neither custom commands recognition. 
There are may be many reasons for that. 
Could you provide a sample project to reproduce the problem?
Also, you could try to remove/comment out the login method that is navigated to. And see if IDE will find the correct login one.

**Comment by MacGyver27** [comment]
Okay - seems the problem was with the Cypress being installed globally.. in our project it needs to be done like that due to CD/CI to avoid constantly downloading it every time there is a run.. 

1) so I did `npm install cypress --no-save` but it still does not recognize the framework
2) when installed it into dev-dependencies it started finally to work `npm install cypress --D`

However now I am getting the command suggestion from the `commands.js` but still am missing then to be highlighted:
![image](https://user-images.githubusercontent.com/6546231/131495452-348c0ed2-28d7-4b08-8efe-cccab2745274.png)
![image](https://user-images.githubusercontent.com/6546231/131495536-d326414f-0b7e-4453-bf30-f19be4463a6c.png)


So my questions are:
1) is there a way to reference global cypress dependency to avoid having it installed locally?
2) fix the custom commands to be highlighted (not white and underscored like in pic)

Thanks

**Comment by mbolotov** [comment]
1. IDE does not index the global libraries, see [this issue](https://youtrack.jetbrains.com/issue/WEB-41931). It seems you created it =). As suggested in this ticket, I tried to create and attach a global variable and found it working.
2. This is a known limitation of the plugin (see [the discussion above](https://github.com/mbolotov/intellij-cypress/issues/50#issuecomment-877174659)). So this functionality is in backlog but I've not yet decided when to implement it. You can add typescript definitions for your custom commands to make the IDE aware of them.

**Comment by MacGyver27** [comment]
1) LOL, true.. completely forgot about it and never checked.
2) Okk, thanks.

**Comment by rnovosad** [comment]
Hi @mbolotov came here from youtube :)
I have a default simple cypress js framework with `index.d.ts` typings for code completion. But after installing a plugin - code completion works only partially. When I go to definition from spec file, WebStorm leads me to the custom command in the support file and from there to `index.d.ts`. From there - I can't find usages of the custom command - but GoTo leads only for the custom command.
I have tried to reproduce the issue on this small project - https://github.com/cypress-io/cypress-example-todomvc.git and it has completely the same issue as mine.
Please have a look

**Comment by mbolotov** [comment]
> to have it fixed in your implementation
Hmm, seems I didn't get this point. Do you mean some functionality is not working after you disabled the option?

**Comment by rnovosad** [comment]
So this is how the autocomplete works without custom commands resolution option
[<img width="557" alt="Screenshot 2021-10-20 at 15 45 13" src="https://user-images.githubusercontent.com/11405694/138096050-39e38530-3a5d-4936-9505-e3228571614d.png">
](url)
And here is with (but with broken typing highlight):
<img width="552" alt="Screenshot 2021-10-20 at 15 46 49" src="https://user-images.githubusercontent.com/11405694/138096155-a60584ca-5ea9-45b2-b83e-08b9a255307d.png">
I'd say it's much cooler with your approach just would be great to have types as all other commands have and to fix the issue with type resolution!
